src/ILVerification/src/TypeVerifier.cs

```diff
@@ -98,36 +99,76 @@ public void VerifyInterfaces()
                         {
-                            VerificationError(VerifierError.InterfaceMethodNotImplemented, Format(type), Format(implementedInterface.DefType), Format(method));
+                            VerificationError(VerifierError.InterfaceMethodNotImplemented, Format(type), Format(implementedInterface.DefType, implementedInterface.InterfaceImplementation), Format(method));
                         }
                     }
                 }
             }
         }
 
-        // Format helpers for future use
-        // The idea is to improve formatting string with tokens value i.e.: 
-        // [Assembly]Class(0x1234) vs current [Assembly]Class
-        // in case somebody will pass '--tokens' switch to ILVerify
         private string Format(EcmaType type)
         {
-            return type.ToString();
+            if (_verifierOptions.IncludeMetadataTokensInErrorMessages)
+            {
+                return string.Format("{0}([{1}]0x{2:X8})", type, _module, _module.MetadataReader.GetToken(_typeDefinitionHandle));
+            }
+            else
+            {
+                return type.ToString();
+            }
         }
 
-        private string Format(DefType interfaceTypeDef)
+        private string Format(DefType defType, InterfaceImplementation interfaceImplementation)
         {
-            return interfaceTypeDef.ToString();
+            if (_verifierOptions.IncludeMetadataTokensInErrorMessages)
+            {
+
+                return string.Format("{0}([{1}]0x{2:X8})", defType, _module, _module.MetadataReader.GetToken(interfaceImplementation.Interface));
+            }
+            else
+            {
+                return defType.ToString();
+            }
         }
 
         private string Format(MethodDesc methodDesc)
         {
-            return methodDesc.ToString();
+            if (_verifierOptions.IncludeMetadataTokensInErrorMessages)
+            {
+                TypeDesc typeDesc = methodDesc.OwningType.GetTypeDefinition();
+                EcmaAssembly ecmaAssembly = (EcmaAssembly)((MetadataType)typeDesc).Module;
+                ILVerifyTypeSystemContext typeSystemContext = (ILVerifyTypeSystemContext)ecmaAssembly.Context;
+                EcmaModule module = typeSystemContext.GetModule(ecmaAssembly.PEReader);
+                MetadataReader reader = ecmaAssembly.PEReader.GetMetadataReader();
+                
+                TypeDefinitionHandle typeDefHandleToken = default;
+                foreach (TypeDefinitionHandle typeDefHandle in reader.TypeDefinitions)
+                {
+                    if (typeDesc == module.GetType(typeDefHandle))
+                    {
+                        typeDefHandleToken = typeDefHandle;
+                        break;
+                    }
+                }
+
+                Debug.Assert(typeDefHandleToken != default);
+
+                return string.Format("{0}([{1}]Interface type token 0x{2:X8}  Method token 0x{3:X8})",
```

Comment:
@jkotas with help of ildasm I understood better where I was doing wrong, thank's!
However I tried another way, I think we need some "mixed" tokens, because if I don't implement a method I don't have a "token" from class that implements. So I improved message in this way:
```
[MD]: Error: Class implements interface but not method, Class: '[InterfaceImplementation]InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented([InterfaceImplementation]0x02000008)' Interface: '[InterfaceDefinition]Interface([InterfaceImplementation]0x01000002)' Missing method: '[InterfaceDefinition]Interface.M3(int32)([InterfaceDefinition]Interface type token 0x02000002  Method token 0x06000003)'.
```
and if I check with ilasm output it's ok, implementers:
class [InterfaceImplementation]0x02000008
implemented interface in current implementer [InterfaceImplementation]0x01000002
```

.class /*02000008*/ public auto ansi beforefieldinit InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented
       extends [System.Runtime/*23000001*/]System.Object/*01000001*/
       implements [InterfaceDefinition/*23000002*/]Interface/*01000002*/
{
  .method /*0600001C*/ public hidebysig newslot virtual final 
          instance void  M1() cil managed
  {
    // Code size       1 (0x1)
    .maxstack  8
    IL_0000:  ret
  } // end of method InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented::M1

  .method /*0600001D*/ public hidebysig newslot virtual final 
          instance void  M2() cil managed
  {
    // Code size       1 (0x1)
    .maxstack  8
    IL_0000:  ret
  } // end of method InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented::M2

  .method /*0600001E*/ public hidebysig newslot virtual final 
          instance void  M3(int64 i) cil managed
  {
    // Code size       1 (0x1)
    .maxstack  8
    IL_0000:  ret
  } // end of method InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented::M3

  .method /*0600001F*/ public hidebysig newslot virtual final 
          instance int32  M4(int32 i) cil managed
  {
    // Code size       2 (0x2)
    .maxstack  8
    IL_0000:  ldc.i4.0
    IL_0001:  ret
  } // end of method InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented::M4
```
Missing method: '[InterfaceDefinition]Interface.M3(int32)([InterfaceDefinition]Interface type token 0x02000002  Method token 0x06000003)'.
```
// =============== CLASS MEMBERS DECLARATION ===================

.class /*02000002*/ interface public abstract auto ansi Interface
{
  .method /*06000001*/ public hidebysig newslot abstract virtual 
          instance void  M1() cil managed
  {
  } // end of method Interface::M1

  .method /*06000002*/ public hidebysig newslot abstract virtual 
          instance void  M2() cil managed
  {
  } // end of method Interface::M2

  .method /*06000003*/ public hidebysig newslot abstract virtual 
          instance void  M3(int32 i) cil managed
  {
  } // end of method Interface::M3

  .method /*06000004*/ public hidebysig newslot abstract virtual 
          instance int32  M4(int32 i) cil managed
  {
  } // end of method Interface::M4

} // end of class Interface
```
Without token:
```
[MD]: Error: Class implements interface but not method, Class: '[InterfaceImplementation]InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented' Interface: '[InterfaceDefinition]Interface' Missing method: '[InterfaceDefinition]Interface.M3(int32)'.
```
thoughts?

Reply:
> `Missing method: '[InterfaceDefinition]Interface.M3(int32)([InterfaceDefinition]Interface type token 0x02000002  Method token 0x06000003)`

I do not think we need to print both type and method tokens. Method token is enough. If somebody needs the type token for the given method token, they can always easily find it using ildasm.

Reply:
Done
```
[MD]: Error: Class implements interface but not method, Class: '[InterfaceImplementation]InvalidParameterTypeM3_InvalidType_InterfaceMethodNotImplemented([InterfaceImplementation]0x02000008)' Interface: '[InterfaceDefinition]Interface([InterfaceImplementation]0x01000002)' Missing method: '[InterfaceDefinition]Interface.M3(int32)([InterfaceDefinition]0x06000003)'.
```
If you agree I left "Missing method" insted of only method, seems more clear to me, but let me know if you don't agree.